File: AutoGenerateSokobanLevel/State.cpp
#include "pch.h"
#include "State.h"
#include <iostream>

State::State(int w, int h)
{
	this->width = w;
	this->height = h;
}

State::~State() {
	delete tiles;
}

void State::setLevel(TileType * tiles)
{
	this->tiles = tiles;
	for (int i = 0; i < this->height; i++) {
		for (int j = 0; j < this->width; j++) {
			if (*(tiles + i * this->width + j) == Character) {
				cx = j;
				cy = i;
			}
		}
	}
}

bool State::ifWin() {
	bool res = true;
	for (int i = 0; i < this->height; i++) {
		for (int j = 0; j < this->width; j++) {
			if (*(tiles + i * this->width + j) == Box) {
				res = false;
			}
		}
	}
	return res;
}
// 向上移动角色
void State::up() {
	int newcx = cx - 1;
	int newcy = cy;
	int newcx2 = cx - 2;
	int newcy2 = cy;
	changLoc(newcx, newcy, newcx2, newcy2);
}
// 向下移动角色
void State::down() {
	int newcx = cx + 1;
	int newcy = cy;
	int newcx2 = cx + 2;
	int newcy2 = cy;
	changLoc(newcx, newcy, newcx2, newcy2);
}
// 向左移动角色
void State::left() {
	int newcx = cx;
	int newcy = cy - 1;
	int newcx2 = cx;
	int newcy2 = cy - 2;
	changLoc(newcx, newcy, newcx2, newcy2);
}
// 向右移动角色
void State::right() {
	int newcx = cx;
	int newcy = cy + 1;
	int newcx2 = cx;
	int newcy2 = cy + 2;
	changLoc(newcx, newcy, newcx2, newcy2);
}

void State::changLoc(int newcx, int newcy, int newcx2, int newcy2) {
	if (*(tiles + newcx * width + newcy) == Wall) {
		return;
	}
	else if (*(tiles + newcx * width + newcy) == Box || *(tiles + newcx * width + newcy) == BoxinAid) {
		if (newcx2 < 0 || newcy2 < 0 || newcx2 >= width || newcy2 >= height || *(tiles + newcx2 * width + newcy2) == Wall || *(tiles + newcx2 * width + newcy2) == Box || *(tiles + newcx2 * width + newcy2) == BoxinAid) {
			return;
		}
		if (*(tiles + newcx * width + newcy) == Box) {
			*(tiles + newcx * width + newcy) = Floor;

		}
		else if (*(tiles + newcx * width + newcy) == BoxinAid) {
			*(tiles + newcx * width + newcy) = Aid;
		}
		if (*(tiles + newcx2 * width + newcy2) == Floor) {
			*(tiles + newcx2 * width + newcy2) = Box;
		}
		else if (*(tiles + newcx2 * width + newcy2) == Aid) {
			*(tiles + newcx2 * width + newcy2) = BoxinAid;
		}
	}
	if (*(tiles + newcx * width + newcy) == Aid) {
		*(tiles + newcx * width + newcy) = CharacterinAid;
	}
	else {
		*(tiles + newcx * width + newcy) = Character;
	}
	if (*(tiles + cx * width + cy) == CharacterinAid) {
		*(tiles + cx * width + cy) = Aid;
	}
	else {
		*(tiles + cx * width + cy) = Floor;
	}
	cx = newcx;
	cy = newcy;
}

State* State::clone() {
	State * newstate = new State(width, height);
	newstate->tiles = new TileType[height * width];
	for (int i = 0; i < this->height; i++) {
		for (int j = 0; j < this->width; j++) {
			newstate->tiles[i * this->width + j] = tiles[i * this->width + j];
		}
	}
	newstate->cx = cx;
	newstate->cy = cy;
	return newstate;
}

bool State::isEqual(State * tempst) {
	for (int i = 0; i < this->height; i++) {
		for (int j = 0; j < this->width; j++) {
			if (tempst->tiles[i * this->width + j] != tiles[i * this->width + j]) {
				return false;
			}
		}
	}
	// std::wcout << "true\n";
	return true;
}
// 判断一个格子人是否能通过
bool State::stepOn(TileType* tt, int i, int  j) {
	if (tt[i * this->width + j] == Floor) {
		tt[i * this->width + j] = Character;
		return true;
	}
	if (tt[i * this->width + j] == Aid) {
		tt[i * this->width + j] = CharacterinAid;
		return true;
	}
	return false;
}
// 利用泛洪算法，标出棋盘上所有角色能够达到的地点。
void State::charFloodFill() {
	while (true) {
		bool ifChange = false;
		for (int i = 1; i < this->height - 1; i++) {
			for (int j = 1; j < this->width - 1; j++) {
				if (tiles[i * this->width + j] == Character || tiles[i*this->width + j] == CharacterinAid) {
					ifChange = ifChange || stepOn(tiles, i - 1, j);
					ifChange = ifChange || stepOn(tiles, i + 1, j);
					ifChange = ifChange || stepOn(tiles, i, j - 1);
					ifChange = ifChange || stepOn(tiles, i, j + 1);
				}
			}
		}
		if (!ifChange) {
			break;
		}
	}
}
// 判断一个箱子能否沿着特定方向被推动，如果能，则返回推动后的状态。
State* State::boxPushed(int i, int j, Direction d) {
	int newi, newj, ci, cj;
	if (d == D_UP) {
		newi = i - 1;
		newj = j;
		ci = i + 1;
		cj = j;
	}
	if (d == D_DOWN) {
		newi = i + 1;
		newj = j;
		ci = i - 1;
		cj = j;
	}
	if (d == D_LEFT) {
		newi = i;
		newj = j - 1;
		ci = i;
		cj = j + 1;
	}
	if (d == D_RIGHT) {
		newi = i;
		newj = j + 1;
		ci = i;
		cj = j - 1;
	}
	if (tiles[ci*width + cj] != Character && tiles[ci*width + cj] != CharacterinAid) {
		return nullptr;
	}
	if (tiles[i*width + j] != Box && tiles[i*width + j] != BoxinAid) {
		return nullptr;
	}
	if (tiles[newi*width + newj] == Wall || tiles[newi*width + newj] == Box || tiles[newi*width + newj] == BoxinAid) {
		return nullptr;
	}
	State * res = clone();
	// 把所有的Character置为Floor，把所有的CharacterInAid置为Aid
	for (int i = 0; i < this->height; i++) {
		for (int j = 0; j < this->width; j++) {
			if (res->tiles[i * this->width + j] == Character) {
				res->tiles[i * this->width + j] = Floor;
			}
			if (res->tiles[i * this->width + j] == CharacterinAid) {
				res->tiles[i * this->width + j] = Aid;
			}
		}
	}
	if (res->tiles[newi*width + newj] == Floor) {
		res->tiles[newi*width + newj] = Box;
	}else if (res->tiles[newi*width + newj] == Aid) {
		res->tiles[newi*width + newj] = BoxinAid;
	}
	if (res->tiles[i*width + j] == Box) {
		res->tiles[i*width + j] = Floor;
	}
	else  if(res->tiles[i*width + j] == BoxinAid){
		res->tiles[i*width + j] = Aid;
	}
	if (res->tiles[i*width + j] == Floor) {
		res->tiles[i*width + j] = Character;
	}
	else  if (res->tiles[i*width + j] == Aid) {
		res->tiles[i*width + j] = CharacterinAid;
	}
	res->cx = j;
	res->cy = i;
	return res;
}

// 剪枝：判断是否死锁
bool State::ifDead() {
	bool res = false;
	res = res || ifWallCorner();
	res = res || ifTwoxTwo();
	return res;
}
// 墙角的死锁
bool State::ifWallCorner() {
	// 第一步，将不可能被进一步推动的BoxinAid置为Wall
	bool ifchange = true;
	while (ifchange) {
		ifchange = false;
		for (int i = 1; i < height - 1; i++) {
			for (int j = 1; j < width - 1; j++) {
				if (tiles[i * width + j] == BoxinAid){
					// 左上
					if (tiles[i * width + j - 1] == Wall && tiles[(i - 1) * width + j] == Wall) {
						tiles[i * width + j] = Wall;
						ifchange = true;
					}
					// 上右
					if (tiles[i * width + j + 1] == Wall && tiles[(i - 1) * width + j] == Wall) {
						tiles[i * width + j] = Wall;
						ifchange = true;
					}
					// 右下
					if (tiles[i * width + j + 1] == Wall && tiles[(i + 1) * width + j] == Wall) {
						tiles[i * width + j] = Wall;
						ifchange = true;
					}
					// 下左
					if (tiles[i * width + j - 1] == Wall && tiles[(i + 1) * width + j] == Wall) {
						tiles[i * width + j] = Wall;
						ifchange = true;
					}
				}
			}
		}
	}
	// 第二步，如果存在不能被进一步被推动的Box，则返回true;
	for (int i = 1; i < height - 1; i++) {
		for (int j = 1; j < width - 1; j++) {
			if (tiles[i * width + j] == Box) {
				// 左上
				if (tiles[i * width + j - 1] == Wall && tiles[(i - 1) * width + j] == Wall) {
					return true;
				}
				// 上右
				if (tiles[i * width + j + 1] == Wall && tiles[(i - 1) * width + j] == Wall) {
					return true;
				}
				// 右下
				if (tiles[i * width + j + 1] == Wall && tiles[(i + 1) * width + j] == Wall) {
					return true;
				}
				// 下左
				if (tiles[i * width + j - 1] == Wall && tiles[(i + 1) * width + j] == Wall) {
					return true;
				}
			}
		}
	}
	return false;
}
// 是否存在四个箱子/墙壁形成一个田字的情况
bool State::ifTwoxTwo() {
	for (int i = 0; i < height - 1; i++) {
		for (int j = 0; j < width - 1; j++) {
			int Boxnum = 0;
			int Wallnum = 0;
			int BoxinAidnum = 0;
			for (int ii = 0; ii < 2; ii++) {
				for (int jj = 0; jj < 2; jj++) {
					if (tiles[(i + ii) * width + j + jj] == Box) {
						Boxnum++;
					}
					if (tiles[(i + ii) * width + j + jj] == Wall) {
						Wallnum++;
					}
					if (tiles[(i + ii) * width + j + jj] == BoxinAid) {
						BoxinAidnum++;
					}
				}
			}
			if ((Boxnum + Wallnum + BoxinAidnum == 4) && Boxnum > 0) {
				return true;
			}
		}
	}
	return false;
}

File: HTML_Sokoban/js/State.js
// State.js - 移植自C++的游戏状态类

import { TileType, Direction } from './GenerateLevel.js';

/**
 * 游戏状态类
 */
class State {
    /**
     * 构造函数
     * @param {number} w - 地图宽度
     * @param {number} h - 地图高度
     */
    constructor(w, h) {
        this.width = w;
        this.height = h;
        this.tiles = null;  // 将在setLevel中初始化
        this.cx = 0;  // 角色X坐标
        this.cy = 0;  // 角色Y坐标
    }

    /**
     * 设置关卡数据
     * @param {Array} tiles - 地图瓦片数组
     */
    setLevel(tiles) {
        this.tiles = tiles;

        // 查找角色位置
        for (let i = 0; i < this.height; i++) {
            for (let j = 0; j < this.width; j++) {
                if (this.tiles[i * this.width + j] === TileType.Character) {
                    this.cx = j;
                    this.cy = i;
                    break;
                }
            }
        }
    }

    /**
     * 判断是否胜利
     * @returns {boolean} 是否胜利
     */
    ifWin() {
        let res = true;
        for (let i = 0; i < this.height; i++) {
            for (let j = 0; j < this.width; j++) {
                if (this.tiles[i * this.width + j] === TileType.Box) {
                    res = false;
                }
            }
        }
        return res;
    }

    /**
     * 向上移动角色
     */
    up() {
        const newcx = this.cx;
        const newcy = this.cy - 1;
        const newcx2 = this.cx;
        const newcy2 = this.cy - 2;
        this.changLoc(newcx, newcy, newcx2, newcy2);
    }

    /**
     * 向下移动角色
     */
    down() {
        const newcx = this.cx;
        const newcy = this.cy + 1;
        const newcx2 = this.cx;
        const newcy2 = this.cy + 2;
        this.changLoc(newcx, newcy, newcx2, newcy2);
    }

    /**
     * 向左移动角色
     */
    left() {
        const newcx = this.cx - 1;
        const newcy = this.cy;
        const newcx2 = this.cx - 2;
        const newcy2 = this.cy;
        this.changLoc(newcx, newcy, newcx2, newcy2);
    }

    /**
     * 向右移动角色
     */
    right() {
        const newcx = this.cx + 1;
        const newcy = this.cy;
        const newcx2 = this.cx + 2;
        const newcy2 = this.cy;
        this.changLoc(newcx, newcy, newcx2, newcy2);
    }

    /**
     * 更改角色位置
     * @param {number} newcx - 新的x坐标
     * @param {number} newcy - 新的y坐标
     * @param {number} newcx2 - 如果推箱子，箱子的新x坐标
     * @param {number} newcy2 - 如果推箱子，箱子的新y坐标
     */
    changLoc(newcx, newcy, newcx2, newcy2) {
        // 如果遇到墙，无法移动
        if (this.tiles[newcy * this.width + newcx] === TileType.Wall) {
            return;
        }
        // 如果遇到箱子，尝试推动它
        else if (this.tiles[newcy * this.width + newcx] === TileType.Box ||
            this.tiles[newcy * this.width + newcx] === TileType.BoxinAid) {
            // 检查箱子能否被推动（边界检查，墙壁检查，箱子检查）
            if (newcx2 < 0 || newcy2 < 0 || newcx2 >= this.width || newcy2 >= this.height ||
                this.tiles[newcy2 * this.width + newcx2] === TileType.Wall ||
                this.tiles[newcy2 * this.width + newcx2] === TileType.Box ||
                this.tiles[newcy2 * this.width + newcx2] === TileType.BoxinAid) {
                return;
            }

            // 处理箱子的移动
            if (this.tiles[newcy * this.width + newcx] === TileType.Box) {
                this.tiles[newcy * this.width + newcx] = TileType.Floor;
            }
            else if (this.tiles[newcy * this.width + newcx] === TileType.BoxinAid) {
                this.tiles[newcy * this.width + newcx] = TileType.Aid;
            }

            // 放置箱子到新位置
            if (this.tiles[newcy2 * this.width + newcx2] === TileType.Floor) {
                this.tiles[newcy2 * this.width + newcx2] = TileType.Box;
            }
            else if (this.tiles[newcy2 * this.width + newcx2] === TileType.Aid) {
                this.tiles[newcy2 * this.width + newcx2] = TileType.BoxinAid;
            }
        }

        // 处理角色的移动
        if (this.tiles[newcy * this.width + newcx] === TileType.Aid) {
            this.tiles[newcy * this.width + newcx] = TileType.CharacterinAid;
        } else {
            this.tiles[newcy * this.width + newcx] = TileType.Character;
        }

        // 处理角色原来位置
        if (this.tiles[this.cy * this.width + this.cx] === TileType.CharacterinAid) {
            this.tiles[this.cy * this.width + this.cx] = TileType.Aid;
        } else {
            this.tiles[this.cy * this.width + this.cx] = TileType.Floor;
        }

        // 更新角色位置
        this.cx = newcx;
        this.cy = newcy;
    }

    /**
     * 克隆当前状态
     * @returns {State} 新状态对象
     */
    clone() {
        const newstate = new State(this.width, this.height);
        newstate.tiles = [...this.tiles];
        newstate.cx = this.cx;
        newstate.cy = this.cy;
        return newstate;
    }

    /**
     * 判断一个状态是否与当前状态相等
     * @param {State} tempst - 比较的状态
     * @returns {boolean} 是否相等
     */
    isEqual(tempst) {
        for (let i = 0; i < this.height; i++) {
            for (let j = 0; j < this.width; j++) {
                if (tempst.tiles[i * this.width + j] !== this.tiles[i * this.width + j]) {
                    return false;
                }
            }
        }
        return true;
    }

    /**
     * 判断一个格子人是否能通过
     * @param {Array} tt - 地图瓦片数组
     * @param {number} i - y坐标
     * @param {number} j - x坐标
     * @returns {boolean} 是否可通过
     */
    stepOn(tt, i, j) {
        if (tt[i * this.width + j] === TileType.Floor) {
            tt[i * this.width + j] = TileType.Character;
            return true;
        }
        if (tt[i * this.width + j] === TileType.Aid) {
            tt[i * this.width + j] = TileType.CharacterinAid;
            return true;
        }
        return false;
    }

    /**
     * 利用泛洪算法，标出棋盘上所有角色能够达到的地点
     */
    charFloodFill() {
        while (true) {
            let ifChange = false;
            for (let i = 1; i < this.height - 1; i++) {
                for (let j = 1; j < this.width - 1; j++) {
                    if (this.tiles[i * this.width + j] === TileType.Character ||
                        this.tiles[i * this.width + j] === TileType.CharacterinAid) {
                        ifChange = ifChange || this.stepOn(this.tiles, i - 1, j);
                        ifChange = ifChange || this.stepOn(this.tiles, i + 1, j);
                        ifChange = ifChange || this.stepOn(this.tiles, i, j - 1);
                        ifChange = ifChange || this.stepOn(this.tiles, i, j + 1);
                    }
                }
            }
            if (!ifChange) {
                break;
            }
        }
    }

    /**
     * 判断一个箱子能否沿着特定方向被推动，如果能，则返回推动后的状态
     * @param {number} i - 箱子的y坐标
     * @param {number} j - 箱子的x坐标
     * @param {Direction} d - 推动方向
     * @returns {State|null} 推动后的状态或null
     */
    boxPushed(i, j, d) {
        let newi, newj, ci, cj;

        // 根据方向计算新位置和角色位置
        if (d === Direction.D_UP) {
            newi = i - 1;
            newj = j;
            ci = i + 1;
            cj = j;
        }
        if (d === Direction.D_DOWN) {
            newi = i + 1;
            newj = j;
            ci = i - 1;
            cj = j;
        }
        if (d === Direction.D_LEFT) {
            newi = i;
            newj = j - 1;
            ci = i;
            cj = j + 1;
        }
        if (d === Direction.D_RIGHT) {
            newi = i;
            newj = j + 1;
            ci = i;
            cj = j - 1;
        }

        // 检查角色是否在箱子对面
        if (this.tiles[ci * this.width + cj] !== TileType.Character &&
            this.tiles[ci * this.width + cj] !== TileType.CharacterinAid) {
            return null;
        }

        // 检查当前位置是否有箱子
        if (this.tiles[i * this.width + j] !== TileType.Box &&
            this.tiles[i * this.width + j] !== TileType.BoxinAid) {
            return null;
        }

        // 检查新位置是否可放置箱子
        if (this.tiles[newi * this.width + newj] === TileType.Wall ||
            this.tiles[newi * this.width + newj] === TileType.Box ||
            this.tiles[newi * this.width + newj] === TileType.BoxinAid) {
            return null;
        }

        // 创建新状态
        let res = this.clone();

        // 先清除所有角色位置
        for (let i = 0; i < this.height; i++) {
            for (let j = 0; j < this.width; j++) {
                if (res.tiles[i * this.width + j] === TileType.Character) {
                    res.tiles[i * this.width + j] = TileType.Floor;
                }
                if (res.tiles[i * this.width + j] === TileType.CharacterinAid) {
                    res.tiles[i * this.width + j] = TileType.Aid;
                }
            }
        }

        // 移动箱子到新位置
        if (res.tiles[newi * this.width + newj] === TileType.Floor) {
            res.tiles[newi * this.width + newj] = TileType.Box;
        } else if (res.tiles[newi * this.width + newj] === TileType.Aid) {
            res.tiles[newi * this.width + newj] = TileType.BoxinAid;
        }

        // 更新原箱子位置
        if (res.tiles[i * this.width + j] === TileType.Box) {
            res.tiles[i * this.width + j] = TileType.Floor;
        } else if (res.tiles[i * this.width + j] === TileType.BoxinAid) {
            res.tiles[i * this.width + j] = TileType.Aid;
        }

        // 放置角色到箱子的原位置
        if (res.tiles[i * this.width + j] === TileType.Floor) {
            res.tiles[i * this.width + j] = TileType.Character;
        } else if (res.tiles[i * this.width + j] === TileType.Aid) {
            res.tiles[i * this.width + j] = TileType.CharacterinAid;
        }

        // 更新角色位置
        res.cx = j;
        res.cy = i;

        return res;
    }

    /**
     * 剪枝：判断是否死锁
     * @returns {boolean} 是否死锁
     */
    ifDead() {
        return this.ifWallCorner() || this.ifTwoxTwo();
    }

    /**
     * 墙角的死锁
     * @returns {boolean} 是否存在墙角死锁
     */
    ifWallCorner() {
        // 第一步，将不可能被进一步推动的BoxinAid置为Wall
        let ifchange = true;

        while (ifchange) {
            ifchange = false;
            for (let i = 1; i < this.height - 1; i++) {
                for (let j = 1; j < this.width - 1; j++) {
                    if (this.tiles[i * this.width + j] === TileType.BoxinAid) {
                        // 左上
                        if (this.tiles[i * this.width + j - 1] === TileType.Wall &&
                            this.tiles[(i - 1) * this.width + j] === TileType.Wall) {
                            this.tiles[i * this.width + j] = TileType.Wall;
                            ifchange = true;
                        }
                        // 上右
                        if (this.tiles[i * this.width + j + 1] === TileType.Wall &&
                            this.tiles[(i - 1) * this.width + j] === TileType.Wall) {
                            this.tiles[i * this.width + j] = TileType.Wall;
                            ifchange = true;
                        }
                        // 右下
                        if (this.tiles[i * this.width + j + 1] === TileType.Wall &&
                            this.tiles[(i + 1) * this.width + j] === TileType.Wall) {
                            this.tiles[i * this.width + j] = TileType.Wall;
                            ifchange = true;
                        }
                        // 下左
                        if (this.tiles[i * this.width + j - 1] === TileType.Wall &&
                            this.tiles[(i + 1) * this.width + j] === TileType.Wall) {
                            this.tiles[i * this.width + j] = TileType.Wall;
                            ifchange = true;
                        }
                    }
                }
            }
        }

        // 第二步，如果存在不能被进一步被推动的Box，则返回true
        for (let i = 1; i < this.height - 1; i++) {
            for (let j = 1; j < this.width - 1; j++) {
                if (this.tiles[i * this.width + j] === TileType.Box) {
                    // 左上
                    if (this.tiles[i * this.width + j - 1] === TileType.Wall &&
                        this.tiles[(i - 1) * this.width + j] === TileType.Wall) {
                        return true;
                    }
                    // 上右
                    if (this.tiles[i * this.width + j + 1] === TileType.Wall &&
                        this.tiles[(i - 1) * this.width + j] === TileType.Wall) {
                        return true;
                    }
                    // 右下
                    if (this.tiles[i * this.width + j + 1] === TileType.Wall &&
                        this.tiles[(i + 1) * this.width + j] === TileType.Wall) {
                        return true;
                    }
                    // 下左
                    if (this.tiles[i * this.width + j - 1] === TileType.Wall &&
                        this.tiles[(i + 1) * this.width + j] === TileType.Wall) {
                        return true;
                    }
                }
            }
        }

        return false;
    }

    /**
     * 是否存在四个箱子/墙壁形成一个田子的情况
     * @returns {boolean} 是否存在田字死锁
     */
    ifTwoxTwo() {
        for (let i = 0; i < this.height - 1; i++) {
            for (let j = 0; j < this.width - 1; j++) {
                let Boxnum = 0;
                let Wallnum = 0;
                let BoxinAidnum = 0;
                let Aidnum = 0;

                for (let ii = 0; ii < 2; ii++) {
                    for (let jj = 0; jj < 2; jj++) {
                        const tile = this.tiles[(i + ii) * this.width + j + jj];
                        if (tile === TileType.Box) {
                            Boxnum++;
                        } else if (tile === TileType.Wall) {
                            Wallnum++;
                        } else if (tile === TileType.BoxinAid) {
                            BoxinAidnum++;
                        } else if (tile === TileType.Aid) {
                            Aidnum++;
                        }
                    }
                }

                // 只有当存在未完成的箱子且形成完全封闭的2x2区域时才是死锁
                // 如果所有箱子都在目标点上，则不是死锁
                if ((Boxnum + Wallnum + BoxinAidnum === 4) && Boxnum > 0) {
                    // 进一步检查：如果这个2x2区域中有足够的目标点，可能不是死锁
                    if (BoxinAidnum + Aidnum < Boxnum) {
                        return true;
                    }
                }
            }
        }

        return false;
    }
}

// 导出模块
export { State }; 